### PR TITLE
(#40) Implement shell editing, loading and saving to config file

### PR DESCRIPTION
**Pull request by catull**
This PR addresses issue #40.
Review is welcome.

### CODE REVIEW

File: src/cm/profile.rs [hunk 1]
```diff
@@ -9,6 +9,7 @@ use std::string::ToString;
 pub struct Profile {
     pub regex_list: StringList,
     pub cmd_list: StringList,
+    pub shell_list: StringList,
```
Comment (rexim) [inline src/cm/profile.rs:12]: Ok, we introduced this shell_list widget, but we never render it anywhere! How are we supposed to use it?

**Review by rexim — CHANGES_REQUESTED** [review]

File: src/cm/profile.rs (hunk 1 above)
Comment (catull) [reply src/cm/profile.rs:11]: You are right, the rendering was not implemented when you reviewed it.

There's an additional commit now which addresses your concern.
Please give it another chance.

File: src/cm/action.rs [hunk 2]
```diff
@@ -66,6 +67,7 @@ pub const NAMES: [&str; LEN] = [
     "jump_to_end",
     "next_search_match",
     "prev_search_match",
+    "edit_shell",
```
Comment (rexim) [inline src/cm/action.rs:70]: What's the point of this action if we edit shells using shell list

File: src/cm/geometry.rs [hunk 3]
```diff
@@ -23,6 +23,12 @@ impl Rect {
                 w: self.w,
                 h: self.h - h,
             },
+            Rect {
+                x: self.x,
+                y: self.y + 2 * h,
+                w: self.w,
+                h: self.h - 2 * h,
+            },
         )
```
Comment (rexim) [inline src/cm/geometry.rs:32]: The layouting of shells list looks really weird to me. I'd expect it to be on the same row with regex and cmd lists.

File: src/cm/profile.rs [hunk 4]
```diff
@@ -46,6 +48,10 @@ impl Profile {
                         cmd_count += 1;
                         result.cmd_list.list.items.push(value.to_string());
                     }
+                    "shells" => {
+                        shell_count += 1;
+                        result.shell_list.list.items.push(value.to_string());
+                    }
```
Comment (rexim) [inline src/cm/profile.rs:54]: Any changes to the default configuration file require migrating old configs. See `migraiton.rs` file.

**Review by rexim — CHANGES_REQUESTED** [review]
On top of my other remarks I feel like this is too complex for what we are trying to achieve. All of these lists mechanism was design for quick and frequent switching between regexes and commands. And I don't think people change their shells that often.

File: src/cm/action.rs (hunk 2 above)
Comment (catull) [reply src/cm/action.rs:70]: True, it is removed now.

File: src/cm/geometry.rs (hunk 3 above)
Comment (catull) [reply src/cm/geometry.rs:32]: Fixed the layouting, now the rect is placed next to the regexes, to the right.

File: src/cm/profile.rs (hunk 4 above)
Comment (catull) [reply src/cm/profile.rs:54]: Provided a migration now, thanks for explaining it.

**Comment by catull** [comment]
Originally, I was thinking of having a single String for the shell, similar to the "user provided command line".
But in order to implement the rendering, I would have to implement the same methods from `string_list.rs` for a that String attribute used in  `profile.st`.

There is no trait that `StringList` implements, yet.
So I would have to introduced it, apply it to `string_list.rs` etc.

That is why I opted for a simpler approach with `shell_list` being of type `StringList`, like `regex_list` and `cmd_list`.

**Side note**:
If you ever look for a different naming, I would suggest `StringList` -> `MultiValueProperty`.
And the shell property would become a `SingleValueProperty`, it could even be a type alias for `String`.

Both would implement a trait named `RenderValueProperty`, which `StringList` already does..

**Review by rexim — APPROVED** [review]

**Comment by rexim** [comment]
@catull looks good to me! :+1: Thank you for your contribution!

**Pull request merged**